app/feature/insurances/src/test/kotlin/com/hedvig/android/feature/insurances/insurance/presentation/InsurancePresenterTest.kt

```diff
@@ -0,0 +1,263 @@
+      id = "crossSellId",
+      title = "crossSellTitle",
+      description = "crossSellDescription",
+      storeUrl = "",
+    ),
+  )
+
+  @Test
+  fun `on launch loading state is true, followed by the loaded state`() = runTest {
+    val getInsuranceContractsUseCase = FakeGetInsuranceContractsUseCase()
+    val getCrossSellsUseCase = FakeGetCrossSellsUseCase()
+    val presenter: InsurancePresenter = InsurancePresenter(
+      getInsuranceContractsUseCase,
+      getCrossSellsUseCase,
+      FakeCrossSellCardNotificationBadgeService(),
+    )
+    presenter.test(InsuranceUiState.InitialState) {
+      awaitItem().also { uiState ->
+        assertThat(uiState).isEqualTo(InsuranceUiState.InitialState)
+        assertThat(uiState.loading).isTrue()
+      }
+
+      getInsuranceContractsUseCase.contracts.add(validContracts)
+      getCrossSellsUseCase.crossSells.add(validCrossSells)
+      awaitItem().also { uiState ->
+        assertAll {
+          assertThat(uiState.hasError).isFalse()
+          assertThat(uiState.loading).isFalse()
+          assertThat(uiState.quantityOfCancelledInsurances)
+            .isEqualTo(validContracts.count(InsuranceContract::isTerminated))
+          assertThat(uiState.insuranceCards.map { it.contractId }).containsSubList(validContracts.map { it.id })
+          assertThat(uiState.crossSells.map { it.id }).containsSubList(validCrossSells.map { it.id })
+          assertThat(uiState.showNotificationBadge)
+        }
+      }
+    }
+  }
+
+  @Test
+  fun `when cross sells fail to load, we get an error state`() = runTest {
+    val getInsuranceContractsUseCase = FakeGetInsuranceContractsUseCase()
+    val getCrossSellsUseCase = FakeGetCrossSellsUseCase()
+    val presenter = InsurancePresenter(
+      getInsuranceContractsUseCase,
+      getCrossSellsUseCase,
+      FakeCrossSellCardNotificationBadgeService(),
+    )
+    presenter.test(InsuranceUiState.InitialState) {
+      skipItems(1)
+
+      getInsuranceContractsUseCase.contracts.add(validContracts)
+      getCrossSellsUseCase.errorMessages.add(ErrorMessage())
+      awaitItem().also { uiState ->
+        assertThat(uiState.hasError).isTrue()
+        assertThat(uiState.loading).isFalse()
+        assertThat(uiState.quantityOfCancelledInsurances).isEqualTo(0)
+      }
+    }
+  }
+
+  @Test
+  fun `partial backend responses keep the state as loading until everything is done`() = runTest {
+    val getInsuranceContractsUseCase = FakeGetInsuranceContractsUseCase()
+    val getCrossSellsUseCase = FakeGetCrossSellsUseCase()
+    val presenter = InsurancePresenter(
+      getInsuranceContractsUseCase,
+      getCrossSellsUseCase,
+      FakeCrossSellCardNotificationBadgeService(),
+    )
+    presenter.test(InsuranceUiState.InitialState) {
+      skipItems(1)
+
+      getInsuranceContractsUseCase.errorMessages.add(ErrorMessage())
+      expectNoEvents() // No events while we're still waiting for all backend calls to finish
```

Comment:
We can also test that after something has happened (in this case, only 1 of the backend responses has arrived) that we've not gotten any UiState updates in this moment in time.